### ISSUE TITLE
Add support for "attributed feature models"

### ISSUE THREAD

**Issue by anb0s** [FeatureIDE/FeatureIDE#255]
For our project we need attributes attached to special features. E.g. it can be some attribute with number range [0..9] or some numbers like [512, 1024, 2048] or some strings ["YES", "NO", "CANCEL"] etc.

Similar approach described here
- P. 2, Figure 1:
  http://gsd.uwaterloo.ca/sites/default/files/Olaechea-NFPinDSML12.pdf
- P.2, Figure 1
  http://cri-dist.univ-paris1.fr/diaz/publications/PL/enase11.pdf

**Comment by tthuem** [comment]
There are some prototypical extensions of FeatureIDE supporting attributes:

* Henneberg's extension https://github.com/tthuem/FeatureIDE/tree/master/experimental/de.ovgu.featureide.fm.extended.ui http://sebastianhenneberg.de/docs/bachelor-thesis.pdf
* SPLConfig http://sourceforge.net/projects/splconfig/ http://homepages.dcc.ufmg.br/~juliana.pereira/papers/2014_CBSoft/SPLConfig.pdf
* CoCo (not yet published)

Another option would be to integrate Clafer within FeatureIDE. However, each of these options requires some engineering effort for a complete integration into FeatureIDE.

**Comment by anb0s** [comment]
Thanks, i want start with "Henneberg's extension":

I've downloaded the sources from master of FeatureIDE repo and tried to compile the "experimental/de.ovgu.featureide.fm.extended.ui" project, but some of files / classes are not there, e.g.:
import de.ovgu.featureide.fm.extended.ui.io.ReaderProblem;
import de.ovgu.featureide.fm.extended.ui.io.attribute.Attributes;

may be i missed something... may be you have already binary distribution of the extended.ui plugin for first test?

**Comment by tthuem** [comment]
We do not have binaries and this extension is no longer maintained. However, you may want to contact: http://sebastianhenneberg.de/

**Comment by anb0s** [comment]
I've contacted Sebastian and he cannot find the whole source code anymore and it will not be maintained, very sad.

I've looked to SPLConfig and it works with model.xml directly (parse etc.) and add some information (one attribute) in own representation (text-file) outside of FeatureIDE context. It's not a generic solution i want to follow. About CoCo i've found only some definitions but no working solution yet. Clafer has own language, IDE and other tools and i think it will be a bunch of work to integrate them to FeatureIDE and it too big (complex) for my usecase.

I need only feature modeling with support of attributes (and may be cardinality in future). Now i will follow two directions:
1. try to understand the not finished solution of Henneberg's extension and add missing parts and integrate it to FeatureIDE (my favourite, but costs lot of my spare time)
2. search for another opensource or commercial tool for feature modeling:
  - pure::variants (commercial) , it looks good but i want to extend it, community version is very basic and has limitations
  - EMF Feature Model (https://www.eclipse.org/featuremodel), the EMF way, looks good, possibility to use state of the art generators (xtend, acceleo etc.), pure::variants and other parties are working on it, may be solution for next yar
  - Yafmt (http://fm-jpikl.blogspot.de), does exactly what i want, difficult to find, no community, feature model with EMF, same possibilities like "EMF Feature Model", needs some polishing (representation of configurations like done in FeatureIDE instead of EMF diagram editor etc.), but it uses good frameworks like EMF, XText, Zest, may be more community support is possible...

**Comment by tthuem** [comment]
We are currently designing new extension point and according interfaces to extend features and feature models in FeatureIDE. So, this could be an opportunity to use them for the attribute extension. If this plan (1.) becomes more concrete, we could discuss whether the new interfaces fit your needs.

**Comment by anb0s** [comment]
Hello,

i'm working now with modified version of Yafmt (added default value handling) for feature models with attributes, but some other models are still done with FeatureIDE. I wrote some external script (ant + xsl) to convert models and configurations from FeatureIDE to Yafmt. I thinking about creating same for Yafmt to FeatureIDE conversion.

Now i want to switch to one tool or contribute missed features to one and waiting for "new extension point and according interfaces to extend features and feature models". I want first verify if this fits my needs and then decide the direction.

I need for my projects :
 - attributes in features (with support for default attribute values defined in model)

 - create constraints to check and modify the attribute values

 - Multi-models and multi-configurations (velvet, MPL)

 - Model and configuration inheritance: base models and base configurations can be inherited and some attribute values can be changed etc.

 - table view to see all (selected) configurations with their features and attributes
background: our actual products database is a Excel sheet with all products in columns and features with attributes in lines, so colleagues want to have similar view to it in one table and select / deselect feature and change attributes in this view

I can write requests and support the development and/or tests if you want. Please give me some "road map" for FeatureIDE.

Thank You!

**Comment by tthuem** [comment]
Markus (and Reimar), could you please comment which of the above requirements are possible with the new feature interfaces? My guess is that all a possible in principle, but some require extensions of FeatureIDE in terms of new Eclipse plug-ins.

**Comment by skrieter** [comment]
Hi,

as for your requests:

> attributes in features (with support for default attribute values defined in model)

This will soon be possible as part of our big refactoring of FeatureIDE, which is currently in development.
Then you can add arbitrary attributes to every feature.

> create constraints to check and modify the attribute values

As the Sat4J boolean solver is a central part of FeatureIDE, it would be very difficult to add constraints over non-boolean attributes.
With our current manpower, this not possible to realize in the near future.

> Multi-models and multi-configurations (velvet, MPL)

There are some implementations for MPL projects that enable multi-feature-modeling, but unfortunately this is currently not in active development.

> Model and configuration inheritance: base models and base configurations can be inherited and some attribute values can be changed etc.

Model inheritance is possible as part of MPL projects and the Velvet language.
However, configuration inheritance is not.
 
> table view to see all (selected) configurations with their features and attributes
background: our actual products database is a Excel sheet with all products in columns and features with attributes in lines, so colleagues want to have similar view to it in one table and select / deselect feature and change attributes in this view

As I already said we currently lack the manpower to implement such greater requests.
However, since this is a separate extension, it is easier to implement than attribute constraints and can possibly be realized as part of a software project in the first quarter of next year.

**Comment by anb0s** [comment]
Hi,

>This will soon be possible as part of our big refactoring of FeatureIDE, which is currently in
> development. Then you can add arbitrary attributes to every feature.

:+1: as soon as you have a stable build, i can try it, thanks!

> As the Sat4J boolean solver is a central part of FeatureIDE, it would be very difficult to add
> constraints over non-boolean attributes. With our current manpower, this not possible to realize in the > near future.

I see, but for attribute usage this is really needed :(. I'm working on prototype for Yafmt, but it's based at XText and grammar definition for such boolean and non-boolean constraint language is possible. So it would not be possible to add own solver for own models with plugin to F-IDE?

>There are some implementations for MPL projects that enable multi-feature-modeling, but unfortunately this is currently not in active development.

Yes, I have filled some requests after my first MPL tests. No idea how to proceed there. So nobody will work at it in near future? P.S.: I saw FAMILIAR project (http://familiar-project.github.io), that has own language for such things. But i think there is no integration for F-IDE available... i will look deeper soon..

> However, since this is a separate extension, it is easier to implement than attribute constraints and can possibly be realized as part of a software project in the first quarter of next year.

For the acceptance in my project i need this "table" presentation, so i can try to start with such a plugin after analyzing the extension points of F-IDE. But it will took some time i think...

Tank You for sharing Feature IDE and keep coding :)

**Comment by tthuem** [comment]
There are some more concrete plans on how to proceed:

1. Enable to add attributes at the feature diagram by means of a dialog. There should be a boolean property to decide whether the value is decided in feature modeling or configuration. Independently, there should be an option to apply/create the property recursively. For configuration-time attributes, it should be possible to specify a default value. Types of attributes should be long, double, and string for now. (Do we need booleans to? Probably not, as they can be encoded as features.)
2. Store those attributes as key-value pairs
3. Create a file in addition to each configuration including all configuration-time attributes and their default values.
4. Extend the configuration editor with a new tab or property view or outline to specify configuration-time attributes.

**Comment by anb0s** [comment]
Thanks for working on this again. My thoughts about the points (from practical project experience):

1. 
> Do we need booleans to?

Yes, i'm using also boolean attributes in other tools too. We have models that have some special meaning coded in attributes, because the feature internal (not extendable) attributes like abstract, etc. are not enough for generating code from models / configurations. The types that i see in most implementations are: boolean, integer (long), float (double) and string. That would be good to have in F-IDE too.
Would be perfect for us, but may be out of scope for F-IDE: We added some more like:
- Enum: just a list with possible literals -> only one can be selected from drop down list by user in model as default and override in configuration; stored as simple list string, format "selected-index-decimal;literal-list", e.g. "0;code_gen_none, code_gen_simple, code_gen_extended"
- BLOB: Byte array with fixed length predefined in model, stored as string, format "length-decimal;hex-bytes-list", e.g. for MAC-Address "6;0x03,0x08,0x00,0x0D,0x0F,0x01"

2. Storing in model
- "model-only-flag" or negated as "overridable-by-configuration"
- type
- name
- default-value

3. ... and overloaded (configured) values too? New readable format for configuration file would be perfect here :) e.g. JSON. All the information can be stored in one file

4. Yes!

**Comment by tthuem** [comment]
@anb0s Thanks for the input. Please drop me a mail (t.thuem@tu-braunschweig.de) that we can discuss your use case before we start coding. :)

**Comment by tthuem** [comment]
@anb0s I still did not get the use case of boolean here. Every feature itself is a boolean. So why not just adding a mandatory feature below instead of boolean flags?

**Comment by anb0s** [comment]
@tthuem The question is if you have one feature with one attribute or one feature with multiple attributes in your "attributed feature model" implementation?

For the first one: yes you can have an optional feature below (selected = true, not selected = false), so may be there is no need for attribute type boolean then.

---
For the second one: i want to define multiple attributes (properties) in one feature and may be some of them are just flags there that should be boolean.

Example from other modeling tool with attributes:
![image](https://user-images.githubusercontent.com/95811/30651564-65fa5c4a-9e26-11e7-92a9-814501ff4faf.png)

config:
![image](https://user-images.githubusercontent.com/95811/30652123-d7176cb4-9e27-11e7-9af6-d9e14f3fb787.png)


Here i have two booleans that have some meaning to one or multiple code generator(s). The target can be source code (C, C++ etc.) or other feature configuration...

**Comment by tthuem** [comment]
In your example, the attributes abstract and generate could also be optional children of Product, right? Does it have any advantages modeling boolean properties as attributes rather than subfeatures?

**Comment by anb0s** [comment]
Yes you are right in classic feature models i would model them as subfeatures and there is no advantage to have boolean properties. But we have some use cases where "feature model" is source for generating different artifacts and we do not want put this information in different models. It means also that some properties are "reused" in more then one feature. In classic way it looks overloaded with lot of subfeatures with similar names, because same name are not possible or with same names if possible. We loose the overview...

What we need here is not a subfeature, but a property (attribute) that can be used in more then one feature. 

P.S.: For the future we need more variant management with support of stereotypes (attribute collections) and constrains for attributes...

**Comment by anb0s** [comment]
see my email to t.thuem@tu-braunschweig.de

**Comment by tthuem** [comment]
Here is what we aim to implement for attributes in the next four weeks:

1. Store and load attributes in XML format for feature models, whereas attributes have
   * name
   * type (i.e., Long, Double, String, Boolean, Enum)
   * unit (e.g., EUR or kg)
   * can be recursive (apply to all subfeatures) or not
   * can be configurable (changes allowed during configuration) or not
   * have a value (even if recursively defined for an ancestor)
2. Create an action and a dialog with a Table Tree View to show and edit attributes
3. Enable filtering for rows based on keywords (e.g., feature names, attribute names, values) that are automatically filled based on the selected feature
4. Create an outline view for configurations showing statistics, such as number of manually/automatically (de)selected and undecided features and a list of those features sorted by name providing navigation to features in editor
5. Compute aggregated values for numerical and boolean attributes shared by at least two features
6. Compute possible ranges for those attributes for the current partial configuration and without any decisions at all

Interestingly, what you mentioned as "reuse" of attributes is what we aim to achieve by letting users define attributes recursively. That is, an recursive attribute is defined for one feature (e.g., the root) and then this attribute is inherited to all its descendants. However, we also aim to define non-recursive attributes.

Thanks for your comments. Indeed, recursive attributes with boolean values are a good example of what would be ugly if one would simply introduce subfeatures, especially in FeatureIDE, as all those would need to have distinguishable names. Hence, I already added boolean values to our planned extension.

**Comment by tthuem** [comment]
Here is a screencast of a prototype, which, however, needs major revision and is far from being stable enough: https://www.youtube.com/watch?v=yhn3avh9U6A&index=4&list=PLBZnGZWZgEBZ_GGTCBQtDMu1KWD7K66O1

Nevertheless, any feedback is welcome.

**Comment by anb0s** [comment]
I've updated my FeatureIDE to 3.5.0 and tried to use extended model and extended configurations with attributes.

First investigations so far:
- it took me some to find how i can use it (new extended model have to be created, did not found menue how to convert the old model to new format)
- added one attribute to model
![image](https://user-images.githubusercontent.com/95811/41193115-8242882c-6c08-11e8-81dd-6b01e921ae07.png)

![image](https://user-images.githubusercontent.com/95811/41193121-8cb32bc2-6c08-11e8-8e17-2d63c4b89602.png)

model.xml:
```
<?xml version="1.0" encoding="UTF-8" standalone="no"?>
<extendedFeatureModel>
    <properties/>
    <struct>
        <and abstract="true" mandatory="true" name="FIDE">
            <feature name="Base">
                <attribute name="value" type="string" unit="" value="hallo"/>
            </feature>
        </and>
    </struct>
    <constraints/>
    <calculations Auto="true" Constraints="true" Features="true" Redundant="true" Tautology="true"/>
    <comments/>
    <featureOrder userDefined="false"/>
</extendedFeatureModel>
```
- created configuration
![image](https://user-images.githubusercontent.com/95811/41193110-75c26892-6c08-11e8-8e56-0053efdb2a5a.png)
file created looks empty:
```
<?xml version="1.0" encoding="UTF-8" standalone="no"?>
<configuration/>
```
- open config --> exception
```
java.lang.NullPointerException
	at de.ovgu.featureide.fm.ui.editors.configuration.ConfigurationEditor.setInput(ConfigurationEditor.java:254)
	at org.eclipse.ui.part.MultiPageEditorPart.init(MultiPageEditorPart.java:784)
	at org.eclipse.ui.internal.EditorReference.initialize(EditorReference.java:362)
	at org.eclipse.ui.internal.e4.compatibility.CompatibilityPart.create(CompatibilityPart.java:342)
	at sun.reflect.GeneratedMethodAccessor67.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(Unknown Source)

```
hm, no idea how to use it...

**Comment by Subaro** [comment]
Hi @anb0s ,

you can find the wiki page [here](https://github.com/FeatureIDE/FeatureIDE/wiki/Feature%20Attributes). It will show you how to create an extended feature model and also how to convert your old model and other things you can do with feature attriubtes.

As for the exception I will need to investigate.

**Comment by anb0s** [comment]
Hi @Subaro 

- still i've not found any information about how to migrate (automatically) old model to new format. I only see description how to create new model
-after creation of new configuration i see the exception
- it looks not usable with 3.5 for me
- is there a special version (nightly other branch etc.) where i can play with it?

Thanks!

**Comment by Subaro** [comment]
@anb0s 

1) Thank you it seems I forgot to describe how to convert a feature model to an extended feature model. To convert a model you need to right-click the feature model and select FeatureIDE -> Export Feature Model. In the FileDialog you need to change the Datatype from 'FeatureIDE *.xml' to 'FeatureIDE (Extended Feature Model) *.xml'. If you convert the feature model 'model.xml' to an extended feature model and override the 'model.xml' be sure to close the FeatureModel Editor and reopen it because overriding open feature model produces some weird things.

2) The configuration error should be fixed. 

3) The extended configuration and the computation of attribute ranges is currenlty only available on the FeatureModel_Attributes branch.

**Comment by tthuem** [comment]
@Subaro Do we still need branch from the team project? If not go ahead and delete it. https://github.com/FeatureIDE/FeatureIDE/tree/tp17_team3

**Comment by Subaro** [comment]
@tthuem Done.

**Comment by tthuem** [comment]
@anb0s Thanks for the report. Could you please attach a minimal example for reproduction purposes?

**Comment by tthuem** [comment]
We were not able to reproduce the above issues without further details.

I'm closing this thread now, as we implemented first support for attributes. We can discuss extensions of this initial support and issues in separate threads then.
